### DISCUSSION TITLE
Duplicate requests to API during SSG and unstable_revalidate

### DISCUSSION THREAD

**Discussion by ksaitor** [vercel/next.js#13591] [Help]
Hi lovely people, Vercel support team suggested me post here.
I've been getting a 2-3 duplicate requests from 2 different IPs a few milliseconds apart.  This is consistent across all routes.

For example:
```
2020-05-28T02:46:58.109745+00:00 app[web.1]: US 3.101.67.38 GET /api/companies
2020-05-28T02:46:58.129604+00:00 app[web.1]: US 54.193.12.162 GET /api/companies
2020-05-28T02:46:58.193972+00:00 app[web.1]: US 3.101.16.223 GET /api/companies
```
Doesn't happen locally - when I run `next dev`. Nor with `next build && next start`.
We fetch data in `getStaticProps` only and are using `axios`. Where could be the issue and how to resolve this?

p.s.
this is causing unnecessary load on the backend + already exhausted my monthly log quote in a few days, causing Heroku go into `Error L10 (output buffer overflow)` mode. Halp!

**Comment by ErnestBrandi** [comment]
I am facing the exact same issue. 

- website is hosted on Vercel
- occurs for such url : /_next/data/[...].json
- use `revalidate` option (could this be the cause ?)
- might cause related APIs to reach their quota's limit quickly

**Reply by ErnestBrandi** [reply]
Scary LogFlare extract 
```
Mon Sep 28 2020 17:51:27 lambda | 200 | *****.fr | /_next/data/Y9nCaZggoH870M7UXoNn1/companies/417555414.json | 92.184.112.253 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1 metadata 
Mon Sep 28 2020 17:51:27 lambda | 200 | *****.fr | /_next/data/Y9nCaZggoH870M7UXoNn1/companies/417555414.json | 92.184.112.253 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1 metadata
Mon Sep 28 2020 17:51:27 lambda | 200 | *****.fr | /_next/data/Y9nCaZggoH870M7UXoNn1/companies/417555414.json | 92.184.112.253 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1 metadata
Mon Sep 28 2020 17:51:27 lambda | 200 | *****.fr | /_next/data/Y9nCaZggoH870M7UXoNn1/companies/309121945.json | 92.184.112.253 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1 metadata
Mon Sep 28 2020 17:51:27 lambda | 200 | *****.fr | /_next/data/Y9nCaZggoH870M7UXoNn1/companies/309121945.json | 92.184.112.253 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1 metadata
Mon Sep 28 2020 17:51:27 lambda | 200 | *****.fr | /_next/data/Y9nCaZggoH870M7UXoNn1/companies/309121945.json | 92.184.112.253 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1 metadata
Mon Sep 28 2020 17:51:27 lambda | 200 | *****.fr | /_next/data/Y9nCaZggoH870M7UXoNn1/companies/810470088.json | 92.184.112.253 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1 metadata
Mon Sep 28 2020 17:51:27 lambda | 200 | *****.fr | /_next/data/Y9nCaZggoH870M7UXoNn1/companies/810470088.json | 92.184.112.253 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1 metadata
Mon Sep 28 2020 17:51:27 lambda | 200 | *****.fr | /_next/data/Y9nCaZggoH870M7UXoNn1/companies/810470088.json | 92.184.112.253 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1 metadata 
```

**Comment by leerob** [comment]
> We fetch data in getStaticProps only and are using axios. 

Two things to consider:

- You shouldn't need to use `axios` with [fetch polyfilled](https://nextjs.org/docs/basic-features/supported-browsers-features#server-side-polyfills)
- You should [not call an API route](https://nextjs.org/docs/basic-features/data-fetching) directly in `getStaticProps`. Instead, directly import the API route and call its function yourself. You may need to slightly refactor your code for this approach.

**Reply by ErnestBrandi** [reply]
What about external API calls ? 
I also use axios in order to benefit its `timeout` and `cancellation` functionalities.

**Reply by leerob** [reply]
External API calls are fine! I was specifically talking about API routes.

**Reply by ErnestBrandi** [reply]
I know, than was more a rhetorical question.
After doing a lot of testing, I found that duplicate logs occurs in a very inconsistent manner (again I host on Vercel) and depend on the build (so this might not be related to Next.js).
With the same commit redeployed, it goes from no log at all to a number of 3 duplicates... (EDIT : actually 4)
I use `getStaticProps` with the `revalidate` option.

**Comment by ksaitor** [comment]
The issue is still there. I've moved away from axios (as an experiment) and I'm still getting double requests for app external API queries while fetching from `getStaticProps`. I've also deployed the same frontend outside of Vercel - tried on Heroku and other containers and the problem is not there - all requests are done only once.

**Comment by mikaelgson** [comment]
I experience something similar to this where I get two serverless executions for one client request. 

From what I can tell, it seems to be because Vercel triggers getStaticProps once to generate the json-file and once to generate the HTML-file. I've verified this by generating a unique id for each execution of getStaticProps and pass it as a prop to the page. Visiting the page directly shows one ID and navigating to it via next-link (using the json-file) will show another.

Why the results of the getStaticProp execution can't be shared to generate both HTML and JSON I'm not sure but maybe it runs on different lambdas as rendering HTML requires more CPU and memory but this is pure speculation.

**Comment by sdsanders** [comment]
I am running into this issue as well. Just to summarize, I'm using `getStaticProps` with the `revalidate` option for incremental static regeneration and hosting on Vercel. Inside of `getStaticProps` I'm making a request to an external API endpoint (via the next.js fetch polyfill) which does a database query. I can see both in Vercel's logs and my API server logs that this request is being made twice - now my database is running twice as many queries as it should be. When I run the same app locally, the request is only made once. @leerob any thoughts?
